makelib/misc.mk

```diff
@@ -0,0 +1,82 @@
+define multi-subst
+$(eval _LIB_MK_TMP_MSUBST_VAL_ := $(strip $3)) \
+$(eval $(foreach s,$1, \
+        $(eval _LIB_MK_TMP_MSUBST_VAL_ := $(subst $s,$2,$(_LIB_MK_TMP_MSUBST_VAL_))))) \
+$(strip $(_LIB_MK_TMP_MSUBST_VAL_)) \
+$(eval _LIB_MK_TMP_MSUBST_VAL_ :=)
+endef
+
+define escape-for-file
+$(strip $(call multi-subst,- / . :,_,$1))
+endef
+
+define path-to-file-with-suffix
+$(strip $(call escape-for-file,$1).$2)
+endef
+
+define stamp-file
+$(STAMPSDIR)/$(call path-to-file-with-suffix,$1,stamp)
+endef
+
+# Generates a stamp filename and assigns it to passed variable
+# name. Generates a stamp's dependency on stamps directory. Adds stamp
+# to CLEAN_FILES. Optional second parameter is for adding a suffix to
+# stamp.
+# Example: $(call setup-custom-stamp-file,FOO_STAMP,/some_suffix)
+define setup-custom-stamp-file
+$(eval $1 := $(call stamp-file,$2)) \
+$(eval $($1): | $$(call to-dir,$($1))) \
+$(eval CLEAN_FILES += $($1))
+endef
+
+# Generates a stamp filename and assigns it to passed variable
+# name. Generates a stamp's dependency on stamps directory. Adds stamp
+# to CLEAN_FILES. Optional second parameter is for adding a suffix to
+# stamp.
+# Example: $(call setup-stamp-file,FOO_STAMP,/some_suffix)
+define setup-stamp-file
+$(eval $(call setup-custom-stamp-file,$1,$(MK_PATH)$2))
+endef
+
+define dep-file
+$(DEPSDIR)/$(call path-to-file-with-suffix,$1,dep.mk)
+endef
+
+define setup-custom-dep-file
+$(eval $1 := $(call dep-file,$2)) \
+$(eval $($1): | $$(call to-dir,$($1))) \
+$(eval CLEAN_FILES += $($1))
+endef
+
+define setup-dep-file
+$(eval $(call setup-custom-dep-file,$1,$(MK_PATH)$2))
+endef
+
+# 1 - test packages
+# 2 - go list format item
+# 3 - excluded directories
+define go-find-directories-unstripped
+$(eval _TST_ESCAPED_REPO_PATH_ := $(REPO_PATH)) \
+$(eval _TST_ESCAPED_REPO_PATH_ := $(subst .,\.,$(_TST_ESCAPED_REPO_PATH_))) \
+$(eval _TST_ESCAPED_REPO_PATH_ := $(subst /,\/,$(_TST_ESCAPED_REPO_PATH_))) \
+$(eval _TST_SPACE_ := ) \
+$(eval _TST_SPACE_ += ) \
+$(eval _TST_FILES_ := $(shell $(GO_ENV) "$(GO)" list -f '{{.ImportPath}} {{.$2}}' $1 | \
+        grep --invert-match '\[\]' | \
+        sed -e 's/.*$(_TST_ESCAPED_REPO_PATH_)\///' -e 's/[[:space:]]*\[.*\]$$//' \
```

Comment:
Before `sed`, the lines are something like:
```
_/home/alban/go/src/github.com/coreos/rkt/common [cgroup_util_test.go]
```
Note that it starts with a "_". It is because `go` does not really support `$GOPATH` with symbolic links going outside of `$GOPATH`, such as:
```
build-rkt-0.5.6+git/gopath/src/github.com/coreos/rkt -> /home/alban/go/src/github.com/coreos/rkt
```
The "_" is usually not a problem because `sed` removed it anyway.

But, when `rkt` is compiled from a directory which does not follow the go directory layout (e.g. Semaphore), it fails. On Semaphore, the lines before `sed` look like:
```
_/home/runner/rkt/builds/build-rkt-coreos-/common [cgroup_util_test.go]
```
So the `grep` on `github.com/coreos/rkt` fails.

Do we want to support compilation when the rkt sources are not in `$GOPATH/src/$REPO_PATH`? It was working before. For Semaphore, this can be worked around by changing `tests/run-build.sh`.